### DISCUSSION TITLE
A better `Excludes` List

### DISCUSSION THREAD

**Discussion by amastis** [pyinstaller/pyinstaller#8319] [PyInstaller]
## The Problem:
Without using UPX, the executable file size grow dramatically based on the packages that are installed in your environment.

## A (limited) solution:
https://github.com/amastis/pyinstaller-excluder

Using a recursive function to find all of the packages that are required (top-level imports from a requirements.txt file), and exclude all other system packages. Some hidden packages may not be properly imported (ex. an Excel handling package used by pandas when doing Excel handling functions in your main script).

## Now unsolved issues become: 
1. How does a user know what some of these hidden imports may be based on what they are doing (and can they find them)?
1. Are there other edge cases that may not be found based on imports?

**Comment by rokm** [comment]
> Without using UPX, the executable file size grow dramatically based on the packages that are installed in your environment.

What is preventing you from creating a minimal virtual environment, with just your dependencies installed in it? That way, optional/conditionally imported packages that you don't really need cannot be pulled in.

> A (limited) solution: [...]

Your "solution" seems to try to populate the excludes list with distribution names, which may or may not correspond to actual importable package names. PyInstaller's excludes list operates on importable names, not distribution names.

> 1. How does a user know what some of these hidden imports may be based on what they are doing (and can they find them)?

Usually by investigating the code of 3rd party package that ends up with a missing module. This requires the user to have at least some experience and/or investigative abilities. Those who lack that end up opening an issue report on our issue tracker.

> 2. Are there other edge cases that may not be found based on imports?

Any python module that is not directly imported using an `ìmport` statement (programmatic imports , lazy loading, etc.) will not be picked up automatically. Same goes for imports made within binary extensions, as we cannot analyze those. And then you also have dynamically loaded shared libraries (loaded from either python code, or from other shared libraries), which also evade our binary dependency analysis. 

But I suspect these have nothing to do with the issue you are trying to solve in the distribution-name space.
